### PR TITLE
Immediate tiering

### PR DESCRIPTION
**Pull request by krisis**
## Description
Support for immediate tiering.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#f23cff2e-055d-43f0-9570-d0c91693ce56) |
| mint-fs.sh | :x: [more...](#8c0f4c7a-1d79-4cb8-8825-9235fd3e441b) |
| mint-erasure.sh | :x: [more...](#6d52bd4b-02b9-42b2-9e55-4e57e64195c8) |
| mint-dist-erasure.sh | :x: [more...](#157a5916-16c2-4ff6-87bb-6372e754fe40) |
| mint-zoned.sh | :x: [more...](#5062720e-1c02-4a56-9f21-6e4e1d116504) |
| mint-gateway-nas.sh | :x: [more...](#fc50e8b9-30c2-403f-8fa5-bd00d837e46d) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#a421277c-1796-4477-a6df-3938d0e63810) |
<a name="f23cff2e-055d-43f0-9570-d0c91693ce56"></a>
### 13298-5820d45/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s1:30218
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 53 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 26 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 2 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 10 seconds
(10/15) Running minio-java tests ... FAILED in 6 seconds
{
  "name": "minio-java",
  "function": "getBucketLifecycle()",
  "duration": 205,
  "status": "FAIL",
  "error": "org.simpleframework.xml.core.ValueRequiredException: Empty value for @org.simpleframework.xml.Element(name=\"StorageClass\", type=void.class, data=false, required=true) on field 'storageClass' private java.lang.String io.minio.messages.NoncurrentVersionTransition.storageClass in class io.minio.messages.NoncurrentVersionTransition at line 1 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.getBucketLifecycle(MinioClient.java:1959), FunctionalTest.getBucketLifecycle(FunctionalTest.java:2869), FunctionalTest.runBucketTests(FunctionalTest.java:3669), FunctionalTest.runTests(FunctionalTest.java:3719), FunctionalTest.main(FunctionalTest.java:3864)]"
}
(10/15) Running minio-js tests ... done in 51 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 24 seconds
(12/15) Running s3cmd tests ... done in 25 seconds
(13/15) Running s3select tests ... done in 4 seconds
(14/15) Running versioning tests ... done in 3 minutes and 5 seconds

Executed 14 out of 15 tests successfully.
```
<a name="6d52bd4b-02b9-42b2-9e55-4e57e64195c8"></a>
### 13298-5820d45/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s6:30098
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 1 minutes and 49 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 20 seconds
(8/15) Running minio-dotnet tests ... done in 47 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 13 seconds
(10/15) Running minio-java tests ... FAILED in 2 seconds
{
  "name": "minio-java",
  "function": "getBucketLifecycle()",
  "duration": 55,
  "status": "FAIL",
  "error": "org.simpleframework.xml.core.ValueRequiredException: Empty value for @org.simpleframework.xml.Element(name=\"StorageClass\", type=void.class, data=false, required=true) on field 'storageClass' private java.lang.String io.minio.messages.NoncurrentVersionTransition.storageClass in class io.minio.messages.NoncurrentVersionTransition at line 1 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.getBucketLifecycle(MinioClient.java:1959), FunctionalTest.getBucketLifecycle(FunctionalTest.java:2869), FunctionalTest.runBucketTests(FunctionalTest.java:3669), FunctionalTest.runTests(FunctionalTest.java:3719), FunctionalTest.main(FunctionalTest.java:3864)]"
}
(10/15) Running minio-js tests ... done in 51 seconds
(11/15) Running minio-py tests ... done in 53 seconds
(12/15) Running s3cmd tests ... done in 24 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running versioning tests ... done in 3 minutes and 2 seconds

Executed 14 out of 15 tests successfully.
```
<a name="fc50e8b9-30c2-403f-8fa5-bd00d837e46d"></a>
### 13298-5820d45/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s8:30022
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 1 minutes and 46 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 17 seconds
(8/15) Running minio-dotnet tests ... done in 36 seconds
(9/15) Running minio-go tests ... done in 40 seconds
(10/15) Running minio-java tests ... FAILED in 2 seconds
{
  "name": "minio-java",
  "function": "getBucketLifecycle()",
  "duration": 42,
  "status": "FAIL",
  "error": "org.simpleframework.xml.core.ValueRequiredException: Empty value for @org.simpleframework.xml.Element(name=\"StorageClass\", type=void.class, data=false, required=true) on field 'storageClass' private java.lang.String io.minio.messages.NoncurrentVersionTransition.storageClass in class io.minio.messages.NoncurrentVersionTransition at line 1 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.getBucketLifecycle(MinioClient.java:1959), FunctionalTest.getBucketLifecycle(FunctionalTest.java:2869), FunctionalTest.runBucketTests(FunctionalTest.java:3669), FunctionalTest.runTests(FunctionalTest.java:3719), FunctionalTest.main(FunctionalTest.java:3864)]"
}
(10/15) Running minio-js tests ... done in 36 seconds
(11/15) Running minio-py tests ... done in 46 seconds
(12/15) Running s3cmd tests ... done in 20 seconds
(13/15) Running s3select tests ... done in 1 seconds
(14/15) Running versioning tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="a421277c-1796-4477-a6df-3938d0e63810"></a>
### 13298-5820d45/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s8:30954
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 1 minutes and 53 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 24 seconds
(8/15) Running minio-dotnet tests ... done in 56 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 5 seconds
(10/15) Running minio-java tests ... FAILED in 6 seconds
{
  "name": "minio-java",
  "function": "getBucketLifecycle()",
  "duration": 168,
  "status": "FAIL",
  "error": "org.simpleframework.xml.core.ValueRequiredException: Empty value for @org.simpleframework.xml.Element(name=\"StorageClass\", type=void.class, data=false, required=true) on field 'storageClass' private java.lang.String io.minio.messages.NoncurrentVersionTransition.storageClass in class io.minio.messages.NoncurrentVersionTransition at line 1 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.getBucketLifecycle(MinioClient.java:1959), FunctionalTest.getBucketLifecycle(FunctionalTest.java:2869), FunctionalTest.runBucketTests(FunctionalTest.java:3669), FunctionalTest.runTests(FunctionalTest.java:3719), FunctionalTest.main(FunctionalTest.java:3864)]"
}
(10/15) Running minio-js tests ... done in 50 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 21 seconds
(12/15) Running s3cmd tests ... done in 25 seconds
(13/15) Running s3select tests ... done in 4 seconds
(14/15) Running versioning tests ... done in 3 minutes and 5 seconds

Executed 14 out of 15 tests successfully.
```
<a name="157a5916-16c2-4ff6-87bb-6372e754fe40"></a>
### 13298-5820d45/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s3:30917
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 1 minutes and 55 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 25 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 1 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 19 seconds
(10/15) Running minio-java tests ... FAILED in 5 seconds
{
  "name": "minio-java",
  "function": "getBucketLifecycle()",
  "duration": 154,
  "status": "FAIL",
  "error": "org.simpleframework.xml.core.ValueRequiredException: Empty value for @org.simpleframework.xml.Element(name=\"StorageClass\", type=void.class, data=false, required=true) on field 'storageClass' private java.lang.String io.minio.messages.NoncurrentVersionTransition.storageClass in class io.minio.messages.NoncurrentVersionTransition at line 1 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.getBucketLifecycle(MinioClient.java:1959), FunctionalTest.getBucketLifecycle(FunctionalTest.java:2869), FunctionalTest.runBucketTests(FunctionalTest.java:3669), FunctionalTest.runTests(FunctionalTest.java:3719), FunctionalTest.main(FunctionalTest.java:3864)]"
}
(10/15) Running minio-js tests ... done in 50 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 18 seconds
(12/15) Running s3cmd tests ... done in 25 seconds
(13/15) Running s3select tests ... done in 4 seconds
(14/15) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 14 out of 15 tests successfully.
```
<a name="5062720e-1c02-4a56-9f21-6e4e1d116504"></a>
### 13298-5820d45/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s8:32686
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 56 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 26 seconds
(8/15) Running minio-dotnet tests ... done in 54 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 41 seconds
(10/15) Running minio-java tests ... FAILED in 6 seconds
{
  "name": "minio-java",
  "function": "getBucketLifecycle()",
  "duration": 208,
  "status": "FAIL",
  "error": "org.simpleframework.xml.core.ValueRequiredException: Empty value for @org.simpleframework.xml.Element(name=\"StorageClass\", type=void.class, data=false, required=true) on field 'storageClass' private java.lang.String io.minio.messages.NoncurrentVersionTransition.storageClass in class io.minio.messages.NoncurrentVersionTransition at line 1 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.getBucketLifecycle(MinioClient.java:1959), FunctionalTest.getBucketLifecycle(FunctionalTest.java:2869), FunctionalTest.runBucketTests(FunctionalTest.java:3669), FunctionalTest.runTests(FunctionalTest.java:3719), FunctionalTest.main(FunctionalTest.java:3864)]"
}
(10/15) Running minio-js tests ... done in 50 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 33 seconds
(12/15) Running s3cmd tests ... done in 25 seconds
(13/15) Running s3select tests ... done in 4 seconds
(14/15) Running versioning tests ... done in 3 minutes and 6 seconds

Executed 14 out of 15 tests successfully.
```
<a name="8c0f4c7a-1d79-4cb8-8825-9235fd3e441b"></a>
### 13298-5820d45/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s8:31721
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 51 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 18 seconds
(8/15) Running minio-dotnet tests ... done in 55 seconds
(9/15) Running minio-go tests ... done in 41 seconds
(10/15) Running minio-java tests ... FAILED in 2 seconds
{
  "name": "minio-java",
  "function": "getBucketLifecycle()",
  "duration": 40,
  "status": "FAIL",
  "error": "org.simpleframework.xml.core.ValueRequiredException: Empty value for @org.simpleframework.xml.Element(name=\"StorageClass\", type=void.class, data=false, required=true) on field 'storageClass' private java.lang.String io.minio.messages.NoncurrentVersionTransition.storageClass in class io.minio.messages.NoncurrentVersionTransition at line 1 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.getBucketLifecycle(MinioClient.java:1959), FunctionalTest.getBucketLifecycle(FunctionalTest.java:2869), FunctionalTest.runBucketTests(FunctionalTest.java:3669), FunctionalTest.runTests(FunctionalTest.java:3719), FunctionalTest.main(FunctionalTest.java:3864)]"
}
(10/15) Running minio-js tests ... done in 46 seconds
(11/15) Running minio-py tests ... done in 38 seconds
(12/15) Running s3cmd tests ... done in 20 seconds
(13/15) Running s3select tests ... done in 1 seconds
(14/15) Running versioning tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/bucket-lifecycle.go [hunk 1]
```diff
@@ -230,6 +230,17 @@ func validateTransitionTier(lc *lifecycle.Lifecycle) error {
 	return nil
 }
 
+// enqueueuTransitionImmediate enqueues obj for transition if eligible.
```
Comment (harshavardhana) [inline cmd/bucket-lifecycle.go]:
```suggestion
// enqueueTransitionImmediate enqueues obj for transition if eligible.
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
We should document this behavior as custom extension in MinIO at `docs/bucket/lifecycle/`

**Comment by klauspost** [comment]
I was actually expecting "immediate" to be blocking, so when a PutObject request returns it *has* been tiered.

I would prefer the data to go directly to tiered storage and never have an immediate stop on the hot tier. 

Other than reducing the IO significantly on the hot tier it would also be an effective rate limiter, though I suspect it will already somewhat be that when the queue is filled.

Is that too complex to implement? Or are there other reasons to keep it like this?

If too complex right now we can of course always revisit it.

File: cmd/object-handlers.go [hunk 2]
```diff
@@ -1851,6 +1856,8 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 
 	// Remove the transitioned object whose object version is being overwritten.
 	if !globalTierConfigMgr.Empty() {
+		// Schedule object for immediate transition if eligible.
+		enqueueTransitionImmediate(objInfo)
 		logger.LogIf(ctx, os.Sweep())
 	}
```
Comment (poornas) [inline cmd/object-handlers.go:1862]: This change is needed in `PutObjectExtractHandler` as well , which incidentally seems to have been missed for the tiering implementation

**Comment by krisis** [comment]
> Is that too complex to implement? Or are there other reasons to keep it like this?
> 
> If too complex right now we can of course always revisit it.

If we tiered objects before we return the PutObject call, the application would observe the performance of the warm-tier. This is why we don't write directly to the warm tier.

**Comment by klauspost** [comment]
@krisis Yeah. I guess that is the tradeoff. Will it block when the queue is full?

**Comment by krisis** [comment]
@klauspost we drop those transition tasks that are enqueued once the queue is full (queue length 10000).  So, these objects won't be immediately transitioned and will be picked up by the scanner eventually.

**Comment by harshavardhana** [comment]
@ravindk89 added you here, we are adding a MinIO specific feature that we perhaps would want to call out for tiering once this is released.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — CHANGES_REQUESTED** [review]
_MINIO_DEBUG_REMOTE_TIERS_IMMEDIATELY - should we be removing this ENV?

**Comment by ravindk89** [comment]
Hm - @krisis is there any functional difference between `_MINIO_DEBUG_REMOTE_TIERS_IMMEDIATELY` and immediate tiering? If not, then agree w/ Harsha on removing.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by krisis** [comment]
> _MINIO_DEBUG_REMOTE_TIERS_IMMEDIATELY - should we be removing this ENV?

I plan to remove it in a separate PR, just to avoid confusing the reviewers. It is a test only feature and can be removed independently.

**Comment by krisis** [comment]
> Hm - @krisis is there any functional difference between `_MINIO_DEBUG_REMOTE_TIERS_IMMEDIATELY` and immediate tiering? If not, then agree w/ Harsha on removing.

There are a few functional differences. With immediate tiering we will not require this test only feature anymore, so it can be removed. I will be removing it in a separate PR. Its presence doesn't affect immediate-tiering feature.

**Comment by harshavardhana** [comment]
> > Hm - @krisis is there any functional difference between `_MINIO_DEBUG_REMOTE_TIERS_IMMEDIATELY` and immediate tiering? If not, then agree w/ Harsha on removing.
> 
> There are a few functional differences. With immediate tiering we will not require this test only feature anymore, so it can be removed. I will be removing it in a separate PR. Its presence doesn't affect immediate-tiering feature.

okay @krisis - I thought it will look cleaner to remove it here. As a self-contained PR i.e this brings in immediate tiering support while it removes these ad-hoc ENVs.

**Review by poornas — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/object-handlers.go (hunk 2 above)
Comment (krisis) [reply cmd/object-handlers.go:1862]: @poornas I think we can add immediate tiering support for PutObjectExtractHandler in a subsequent PR.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/object-handlers.go (hunk 2 above)
Comment (poornas) [reply cmd/object-handlers.go:1862]: okay

**Review by poornas — APPROVED** [review]
LGTM

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**